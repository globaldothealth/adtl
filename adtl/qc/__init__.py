"""
Quality Control module for ADTL
"""
import math
import json
import functools
import copy
from pathlib import Path
from typing import List, TypedDict, Optional, Callable, Tuple, Dict, Any

import pandas as pd
import numpy as np
import fastjsonschema


class Rule(TypedDict):
    module: str
    name: str
    description: str
    long_description: Optional[str]
    pattern: str


class WorkUnit(TypedDict):
    rule: Rule
    file: str
    dataset: str


class Dataset(TypedDict):
    folder: str
    files: List[str]


class WorkUnitResult(TypedDict):
    rule: str
    dataset: str
    file: str
    rows_success: int
    rows_fail: int
    rows: int
    ratio_success: float
    rows_fail_idx: List[int]
    success: bool
    mostly: float
    reason: str
    fail_data: pd.DataFrame


def rules_for(pattern: str, *rules):
    for r in rules:
        r.pattern = pattern


def rule(columns: List[str], mostly: float = 0, set_missing_columns: bool = True):
    """Decorator that indicates a QC rule

    Args:
        columns: List of required columns in the dataframe. If columns are not
        present, they are assigned null.

        mostly: Indicates the check in the rule should be considered a success
        if the ratio of successful rows is greater than this value. If not
        specified all rows have to succeed for the rule to be a considered a
        success.

        set_missing_columns: On by default, this setting assigns NA/null to
        required columns that are missing in the data
    """

    def decorator_rule(func):
        @functools.wraps(func)
        def wrapper(df, **kwargs):
            if set_missing_columns:
                for c in set(columns) - set(df.columns):
                    df[c] = None
            series = func(df, **kwargs)
            assert len(series) == len(
                df
            ), "Returned series must have same cardinality as source dataframe"
            rows_fail_idx = [i for i, val in enumerate(series) if val is False]
            if isinstance(series, (pd.Series, np.ndarray)):
                rows_success: int = series.sum()
                rows_fail = len(series) - rows_success
                ratio_success = rows_success / len(series)
                return dict(
                    rows_success=int(rows_success),
                    rows_fail=int(rows_fail),
                    rows=int(rows_success) + int(rows_fail),
                    ratio_success=ratio_success,
                    success=bool(ratio_success >= mostly),
                    mostly=mostly,
                    rows_fail_idx=",".join(map(str, rows_fail_idx)),
                    fail_data=json.dumps(
                        df.loc[rows_fail_idx][columns].to_dict(),
                        sort_keys=True,
                    ),
                )
            elif isinstance(series, bool):
                return dict(
                    rows_success=None,
                    rows_fail=None,
                    rows=None,
                    ratio_success=None,
                    success=series,
                    mostly=mostly,
                    rows_fail_idx=None,
                    fail_data=None,
                )

        return wrapper

    return decorator_rule


def _get_columns_from_valid_data(df: pd.DataFrame) -> List[str]:
    cols = set()
    for row in df.itertuples():
        cols |= set(row.column.split(";"))
    return sorted(cols - {""})


def _to_json(row: Dict[str, Any]) -> Dict[str, Any]:
    """"""
    nrow = copy.deepcopy(row)
    for key, val in row.items():
        if isinstance(val, int):
            continue
        elif isinstance(val, float):
            if math.isnan(val):
                del nrow[key]
            continue
        else:
            val = str(val).strip()
        if val.upper() == "TRUE":
            nrow[key] = True
            continue
        elif val.upper() == "FALSE":
            nrow[key] = False
            continue
        elif val.upper() in ["NA", "NK", "N/A", "NAN", "N/K", "NONE", "-", ""]:
            del nrow[key]
            continue
        else:
            pass
        try:
            nrow[key] = int(val)
        except ValueError:
            try:
                nrow[key] = float(val)
            except ValueError:
                nrow[key] = str(val)
    return nrow


def schema(
    schema_path: str, pattern: str = "*.csv", mostly: float = 0.95
) -> Callable[[pd.DataFrame], List[WorkUnitResult]]:
    schema_path = Path(schema_path)  # type: ignore
    with schema_path.open() as fp:
        schema = json.load(fp)
        validator = fastjsonschema.compile(schema)

    def rule_schema(df: pd.DataFrame):
        valids: List[Tuple[bool, str, str]] = []
        for row in map(_to_json, df.to_dict(orient="records")):
            try:
                validator(row)
                valids.append((True, "", ""))
            except fastjsonschema.exceptions.JsonSchemaValueException as e:
                valids.append((False, e.message, ";".join(e.path[1:])))
        valid_data = pd.DataFrame(valids, columns=["is_valid", "reason", "column"])
        rows_success = valid_data.is_valid.sum()
        rows_fail = len(valid_data) - rows_success
        ratio_success = rows_success / (rows_success + rows_fail)
        reason_counts = valid_data[~valid_data.is_valid].reason.value_counts()
        res = [
            dict(
                rows_success=rows_success,
                rows_fail=rows_fail,
                rows=rows_success + rows_fail,
                ratio_success=ratio_success,
                success=bool(ratio_success >= mostly),
                mostly=mostly,
                rows_fail_idx=[i for i, v in enumerate(valids) if v[0] is False],
                fail_data=None,
            )
        ]
        for reason, count in zip(reason_counts.index, reason_counts):
            res.append(
                dict(
                    rows=count,
                    rows_success=0,
                    rows_fail=count,
                    ratio_success=0,
                    success=False,
                    mostly=0,
                    rows_fail_idx=list(
                        valid_data.loc[valid_data.reason == reason].index
                    ),
                    reason=reason,
                    fail_data=json.dumps(
                        df.loc[valid_data.reason == reason][
                            _get_columns_from_valid_data(
                                valid_data[valid_data.reason == reason]
                            )
                        ]
                        .head(10)
                        .to_dict()
                    ),
                )
            )
        return res

    rule_schema.__doc__ = f"{schema.get('title', schema_path)} schema"
    rule_schema.__name__ = "schema_" + schema_path.stem.split(".")[0]
    rule_schema.pattern = pattern
    return rule_schema


def main(args=None):
    from .runner import _main

    _main(args)
