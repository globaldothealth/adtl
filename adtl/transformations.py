""" Functions which can be applied to source fields, allowing extensibility """

import logging
from datetime import datetime, timedelta, date

try:
    import zoneinfo
except ImportError:
    from backports import zoneinfo  # noqa

import pint


def isNotNull(value):
    return value not in [None, ""]


def textIfNotNull(field, return_val):
    return return_val if field not in [None, ""] else None


def getFloat(value):
    if not value:
        return None

    if '"' in value or " " in value:
        value = value.strip('"').replace(" ", "")

    # handle comma decimal separator
    value_int, _, fraction = value.partition(",")
    if "." in fraction:  # comma was being used as a thousands separator
        value = value_int + fraction
    else:
        # replace full stops as they may be used for thousands separator, first
        # then use full stop as decimal separator
        value = value.replace(".", "").replace(",", ".")
    try:
        return float(value)
    except ValueError:
        return value


def yearsElapsed(birthdate, currentdate):
    if birthdate in [None, ""] or currentdate in [None, ""]:
        return None

    bd = datetime.strptime(birthdate, "%Y-%m-%d")
    cd = datetime.strptime(currentdate, "%Y-%m-%d")

    days = cd - bd
    return pint.Quantity(days.days, "days").to("years").m


def durationDays(symptomstartdate, currentdate):
    """
    Returns the number of days between two dates.
    Preferable to Y-M-D elapsed, as month length is ambiguous -
    can be anywhere between 28-31 days.
    """
    if symptomstartdate in [None, ""] or currentdate in [None, ""]:
        return None

    bd = datetime.strptime(symptomstartdate, "%Y-%m-%d")
    cd = datetime.strptime(currentdate, "%Y-%m-%d")

    days = cd - bd
    return days.days


def startDate(enddate, duration):
    """
    Retuns the start date in ISO format, given the end date and the duration.
    """
    if enddate in [None, ""] or duration in [None, ""]:
        return None

    ed = datetime.strptime(enddate, "%Y-%m-%d")

    sd = ed - timedelta(days=duration)

    return sd.strftime("%Y-%m-%d")


def endDate(startdate, duration):
    """
    Retuns the end date in ISO format, given the start date and the duration.
    """
    if startdate in [None, ""] or duration in [None, ""]:
        return None

    sd = datetime.strptime(startdate, "%Y-%m-%d")

    duration = float(duration)

    ed = sd + timedelta(days=duration)

    return ed.strftime("%Y-%m-%d")


def makeDate(year, month, day):
    if year in [None, ""] or month in [None, ""] or day in [None, ""]:
        return None
    try:
        year, month, day = int(year), int(month), int(day)
    except ValueError:
        logging.error(
            f"Error in casting to integer: year={year}, month={month}, day={day}"
        )
        return None
    try:
        return date(year, month, day).isoformat()
    except ValueError:
        logging.error(
            f"Could not construct date from: year={year}, month={month}, day={day}"
        )
        return None


def makeDateTimeFromSeconds(date, time_seconds, date_format, tzname):
    if date == "":
        return None
    try:
        t = datetime.strptime(date, date_format).replace(
            tzinfo=zoneinfo.ZoneInfo(tzname)
        )
    except ValueError:
        logging.error(
            f"Could not convert date {date!r} from date format {date_format!r}"
        )
        return None
    if time_seconds == "":
        return t.date().isoformat()  # return date only
    time_seconds = int(time_seconds)
    hour = time_seconds // 3600
    minute = (time_seconds % 3600) // 60
    return t.replace(hour=hour, minute=minute).isoformat()
