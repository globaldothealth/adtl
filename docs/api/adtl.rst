=====================
Module reference
=====================

.. autoclass:: adtl.Parser
  :members:
