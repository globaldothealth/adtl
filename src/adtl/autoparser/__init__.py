import argparse
import sys

try:
    from .dict_writer import DictWriter, create_dict, generate_descriptions
    from .make_toml import ParserGenerator, create_parser
    from .mapping import Mapper, create_mapping
except ImportError:  # pragma: no cover
    raise ImportError(
        "autoparser is not available. Import as 'adtl[autoparser]' to use."
    )

__all__ = [
    "DictWriter",
    "create_dict",
    "generate_descriptions",
    "Mapper",
    "create_mapping",
    "ParserGenerator",
    "create_parser",
]

from .dict_writer import main as make_dd_main
from .make_toml import main as make_toml_main
from .mapping import main as csv_mapping_main


def main():
    parser = argparse.ArgumentParser(
        description="adtl-autoparser: A tool for creating data dictionaries and parsers"
    )

    subparsers = parser.add_subparsers(title="subcommands", dest="subcommand")

    # Subcommand: create-dict
    parser_create_dict = subparsers.add_parser(
        "create-dict",
        help="Create a data dictionary from a dataset",
    )
    parser_create_dict.set_defaults(func=make_dd_main)

    # Subcommand: create-mapping
    parser_create_mapping = subparsers.add_parser(
        "create-mapping",
        help="Create initial CSV mapping from data dictionary (LLM key required)",
    )
    parser_create_mapping.set_defaults(func=csv_mapping_main)

    # Subcommand: create-parser
    parser_create_parser = subparsers.add_parser(
        "create-parser",
        help="Generate TOML parser from CSV mapping file",
    )
    parser_create_parser.set_defaults(func=make_toml_main)

    args, unknown_args = parser.parse_known_args()

    if args.subcommand is None:
        parser.print_help()
        sys.exit(1)

    # Call the appropriate function with remaining arguments
    args.func(unknown_args)
