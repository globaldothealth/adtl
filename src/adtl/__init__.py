import argparse
import importlib.metadata
import json

from adtl.parser import Parser
from adtl.python_interface import parse

__all__ = ["Parser", "parse"]

__version__ = importlib.metadata.version("adtl")


def main(argv=None):
    cmd = argparse.ArgumentParser(
        prog="adtl",
        description="Transforms and validates data into CSV given a specification",
    )
    cmd.add_argument(
        "spec",
        help="specification file to use",
    )
    cmd.add_argument("file", help="file to read in")
    cmd.add_argument(
        "-o", "--output", help="output file, if blank, writes to standard output"
    )
    cmd.add_argument(
        "--encoding", help="encoding input file is in", default="utf-8-sig"
    )
    cmd.add_argument(
        "--parquet", help="output file is in parquet format", action="store_true"
    )
    cmd.add_argument(
        "-q",
        "--quiet",
        help="quiet mode - decrease verbosity, disable progress bar",
        action="store_true",
    )
    cmd.add_argument("--save-report", help="save report in JSON format")
    cmd.add_argument(
        "--include-def",
        action="append",
        help="include external definition (TOML or JSON)",
    )
    cmd.add_argument("--version", action="version", version="%(prog)s " + __version__)
    args = cmd.parse_args(argv)
    include_defs = args.include_def or []
    spec = Parser(args.spec, include_defs=include_defs, quiet=args.quiet)

    # check for incompatible options
    if spec.header.get("returnUnmatched") and args.parquet:
        raise ValueError("returnUnmatched and parquet options are incompatible")

    # run adtl
    adtl_output = spec.parse(args.file, encoding=args.encoding)
    adtl_output.save(args.output or spec.name, "parquet" if args.parquet else "csv")
    if args.save_report:
        adtl_output.report.update(
            dict(
                encoding=args.encoding,
                include_defs=include_defs,
                file=args.file,
                parser=args.spec,
            )
        )
        with open(args.save_report, "w") as fp:
            json.dump(adtl_output.report, fp, sort_keys=True, indent=2)
    else:
        adtl_output.show_report()


if __name__ == "__main__":
    main()
