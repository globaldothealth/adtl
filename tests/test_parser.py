import sys
import json
from pathlib import Path
import pytest

import adtl as parser

RULE_SINGLE_FIELD = {"field": "diabetes_mhyn"}
RULE_SINGLE_FIELD_WITH_MAPPING = {
    "field": "diabetes_mhyn",
    "values": {"1": True, "2": False, "3": None},
}

TEST_PARSERS_PATH = Path(__file__).parent / "parsers"
TEST_SOURCES_PATH = Path(__file__).parent / "sources"


LIVER_DISEASE = [
    {
        "field": "modliv",
        "values": {"1": True, "0": False, "2": None},
        "description": "Moderate liver disease",
    },
    {
        "field": "mildliver",
        "values": {"1": True, "0": False, "2": None},
        "description": "Mild liver disease",
    },
]

ROW_CONCISE = {"mildliv": 0, "modliv": 2}
RULE_EXCLUDE = {
    "combinedType": "list",
    "fields": [{"field": "mildliv"}, {"field": "modliv"}],
}

ROW_CONDITIONAL = {"outcome_date": "2022-01-01", "outcome_type": 4}
RULE_CONDITIONAL_OK = {"field": "outcome_date", "if": {"outcome_type": 4}}
RULE_CONDITIONAL_FAIL = {"field": "outcome_date", "if": {"outcome_type": {"<": 4}}}

ROW_UNIT_MONTH = {"age": 18, "age_unit": "1"}
ROW_UNIT_YEAR = {"age": 18, "age_unit": "2"}
RULE_UNIT = {
    "field": "age",
    "unit": "years",
    "source_unit": {"field": "age_unit", "values": {"1": "months", "2": "years"}},
}

RULE_COMBINED_TYPE_ANY = {"combinedType": "any", "fields": LIVER_DISEASE}
RULE_COMBINED_TYPE_ALL = {"combinedType": "all", "fields": LIVER_DISEASE}
RULE_COMBINED_FIRST_NON_NULL = {
    "combinedType": "firstNonNull",
    "fields": [{"field": "first"}, {"field": "second"}],
}
RULE_COMBINED_TYPE_LIST = {"combinedType": "list", "fields": LIVER_DISEASE}
RULE_COMBINED_TYPE_LIST_PATTERN = {
    "combinedType": "list",
    "fields": [
        {"fieldPattern": ".*liv.*", "values": {"1": True, "0": False, "2": None}}
    ],
}

RULE_NON_SENSITIVE = {"field": "id"}
RULE_SENSITIVE = {"field": "id", "sensitive": True}

RULE_DATE_MDY = {"field": "outcome_date", "source_date": "%d/%m/%Y", "date": "%m/%d/%Y"}
RULE_DATE_ISO = {"field": "outcome_date", "source_date": "%d/%m/%Y"}

ONE_MANY_SOURCE = [
    {"dt": "2022-02-05", "headache_cmyn": 1, "cough_cmyn": 1, "dyspnea_cmyn": 0}
]

ONE_MANY_OUTPUT = [
    {"date": "2022-02-05", "name": "headache", "is_present": True},
    {"date": "2022-02-05", "name": "cough", "is_present": True},
]

SOURCE_GROUPBY = [
    {"sex": "1", "subjid": "007", "dsstdat": "2020-05-06", "hostdat": "2020-06-08"},
    {"sex": "2", "subjid": "001", "dsstdat": "2022-01-11", "hostdat": "2020-06-08"},
]

BUFFER_GROUPBY = """
sex_at_birth,subject_id,dataset_id,country_iso3,enrolment_date,admission_date
male,007,dataset-2020-03-23,GBR,2020-05-06,2020-06-08
female,001,dataset-2020-03-23,GBR,2022-01-11,2020-06-08
"""


@pytest.mark.parametrize(
    "row_rule,expected",
    [
        (({"diabetes_mhyn": "1"}, RULE_SINGLE_FIELD_WITH_MAPPING), True),
        (({"diabetes_mhyn": "1"}, RULE_SINGLE_FIELD), "1"),
        (({}, "CONST"), "CONST"),
        (({"modliv": "1", "mildliver": "0"}, RULE_COMBINED_TYPE_ANY), True),
        (({"modliv": "1", "mildliver": "0"}, RULE_COMBINED_TYPE_ALL), False),
        (({"modliv": "1", "mildliver": "0"}, RULE_COMBINED_TYPE_LIST), [True, False]),
        (
            ({"modliv": "1", "mildliver": "0"}, RULE_COMBINED_TYPE_LIST_PATTERN),
            [True, False],
        ),
        (
            (
                {"modliv": "1", "mildliver": "3"},
                {**RULE_COMBINED_TYPE_LIST_PATTERN, "excludeWhen": None},
            ),
            [True],
        ),
        (
            ({"modliv": "1", "mildliver": "3"}, RULE_COMBINED_TYPE_LIST_PATTERN),
            [True, None],
        ),
        (({"id": "1"}, RULE_NON_SENSITIVE), "1"),
        (
            ({"id": "1"}, RULE_SENSITIVE),
            "6b86b273ff34fce19d6b804eff5a3f5747ada4eaa22f1d49c01e52ddb7875b4b",
        ),
        (({"first": "1", "second": ""}, RULE_COMBINED_FIRST_NON_NULL), "1"),
        (({"first": "1", "second": "2"}, RULE_COMBINED_FIRST_NON_NULL), "1"),
        (({"first": "2", "second": "1"}, RULE_COMBINED_FIRST_NON_NULL), "2"),
        (({"first": "", "second": "3"}, RULE_COMBINED_FIRST_NON_NULL), "3"),
        ((ROW_CONDITIONAL, RULE_CONDITIONAL_OK), "2022-01-01"),
        ((ROW_CONDITIONAL, RULE_CONDITIONAL_FAIL), None),
        ((ROW_UNIT_MONTH, RULE_UNIT), 1.5),
        ((ROW_UNIT_YEAR, RULE_UNIT), 18),
        (({"outcome_date": "02/05/2022"}, RULE_DATE_MDY), "05/02/2022"),
        (({"outcome_date": "02/05/2022"}, RULE_DATE_ISO), "2022-05-02"),
    ],
)
def test_get_value(row_rule, expected):
    row, rule = row_rule
    assert parser.get_value(row, rule) == expected


@pytest.mark.parametrize(
    "row_rule,expected",
    [
        ((ROW_CONDITIONAL, {"outcome_type": 4}), True),
        ((ROW_CONDITIONAL, {"outcome_type": {"==": 4}}), True),
        ((ROW_CONDITIONAL, {"outcome_type": 3}), False),
        ((ROW_CONDITIONAL, {"outcome_type": {">": 2}}), True),
        ((ROW_CONDITIONAL, {"outcome_type": {"<": 10}}), True),
        ((ROW_CONDITIONAL, {"outcome_type": {"<=": 4}}), True),
        ((ROW_CONDITIONAL, {"outcome_type": {">=": 4}}), True),
        ((ROW_CONDITIONAL, {"outcome_type": {"<": 10}}), True),
        ((ROW_CONDITIONAL, {"outcome_type": {"!=": 4}}), False),
        (
            (
                ROW_CONDITIONAL,
                {"any": [{"outcome_type": {">": 2}}, {"outcome_date": {"<": "2022"}}]},
            ),
            True,
        ),
        (
            (
                ROW_CONDITIONAL,
                {"all": [{"outcome_type": {">": 2}}, {"outcome_date": {"<": "2022"}}]},
            ),
            False,
        ),
    ],
)
def test_parse_if(row_rule, expected):
    assert parser.parse_if(*row_rule) == expected


def test_one_to_many():
    actual_one_many_output_rows = list(
        parser.Parser(TEST_PARSERS_PATH / "oneToMany.json")
        .parse_rows(ONE_MANY_SOURCE)
        .read_table("observation")
    )
    actual_one_many_output_csv = list(
        parser.Parser(TEST_PARSERS_PATH / "oneToMany.json")
        .parse(TEST_SOURCES_PATH / "oneToMany.csv")
        .read_table("observation")
    )
    assert actual_one_many_output_rows == ONE_MANY_OUTPUT
    assert actual_one_many_output_csv == ONE_MANY_OUTPUT


# test exceptions


def test_exception_get_value_unhashed():
    with pytest.raises(ValueError, match="Could not convert"):
        parser.get_value_unhashed(
            {"age_unit": "years", "age": "a"},
            {"source_unit": {"field": "age_unit"}, "field": "age", "unit": "months"},
        )


def test_invalid_rule():
    with pytest.raises(ValueError, match="Could not return value for"):
        parser.get_value_unhashed({"age_unit": "years", "age": "a"}, {})


def test_invalid_operand_parse_if():
    with pytest.raises(ValueError, match="Unrecognized operand"):
        parser.parse_if(
            {"outcome_type": 1, "outcome_date": "2022-06-04"},
            {"outcome_type": {"<>": 5}},
        )
    with pytest.raises(ValueError, match="if-subexpressions should be a dictionary"):
        parser.parse_if(
            {"outcome_type": 1, "outcome_date": "2022-06-04"},
            {"outcome_type": {"<>", 5}},
        )


@pytest.mark.parametrize(
    "rowrule,expected",
    [
        ((ROW_CONCISE, RULE_EXCLUDE), [0, 2]),
        ((ROW_CONCISE, {**RULE_EXCLUDE, "excludeWhen": False}), [2]),
        ((ROW_CONCISE, {**RULE_EXCLUDE, "excludeWhen": None}), [0, 2]),
        ((ROW_CONCISE, {**RULE_EXCLUDE, "excludeWhen": [2]}), [0]),
    ],
)
def test_list_exclude(rowrule, expected):
    assert parser.get_combined_type(*rowrule) == expected


def test_invalid_list_exclude():
    with pytest.raises(
        ValueError, match="excludeWhen rule should be null, false, or a list of values"
    ):
        parser.get_combined_type(
            {"modliv": 1, "mildliv": 2},
            {
                "combinedType": "list",
                "fields": [{"field": "modliv"}, {"field": "mildliv"}],
                "excludeWhen": 5,
            },
        )


def test_invalid_combined_type():
    with pytest.raises(ValueError, match="Unknown"):
        parser.get_combined_type(ROW_CONCISE, {"combinedType": "collage", "fields": []})


def test_validate_spec():
    with pytest.raises(ValueError, match="Specification header requires key"):
        ps = parser.Parser(dict())


@pytest.mark.parametrize(
    "source,expected",
    [
        (TEST_PARSERS_PATH / "oneToMany.json", ["observation"]),
        (TEST_PARSERS_PATH / "groupBy.json", ["subject"]),
    ],
)
def test_load_spec(source, expected):
    ps = parser.Parser(source)
    assert list(ps.tables.keys()) == expected


def test_parse_write_buffer(snapshot):
    ps = parser.Parser(TEST_PARSERS_PATH / "groupBy.json")
    buf = ps.parse_rows(SOURCE_GROUPBY).write_csv("subject")
    assert buf == snapshot


@pytest.mark.parametrize(
    "source,error",
    [
        (
            TEST_PARSERS_PATH / "groupBy-missing-kind.json",
            "Required 'kind' attribute within 'tables' not present for",
        ),
        (
            TEST_PARSERS_PATH / "groupBy-missing-table.json",
            "Parser specification missing required",
        ),
        (
            TEST_PARSERS_PATH / "groupBy-incorrect-aggregation.json",
            "groupBy needs aggregation=lastNotNull to be set for table:",
        ),
    ],
)
def test_invalid_spec_raises_error(source, error):
    with pytest.raises(ValueError, match=error):
        ps = parser.Parser(source)


def test_parser_clear():
    ps = parser.Parser(TEST_PARSERS_PATH / "oneToMany.json")
    ps.data = {"observation": []}
    ps.clear()
    assert ps.data == {}


def test_read_table_raises_error():
    with pytest.raises(ValueError, match="Invalid table"):
        list(
            parser.Parser(TEST_PARSERS_PATH / "oneToMany.json")
            .parse_rows(ONE_MANY_SOURCE)
            .read_table("obs")
        )


def test_constant_table():
    ps = parser.Parser(TEST_PARSERS_PATH / "constant.json").parse_rows([{"x": 1}])
    assert list(ps.read_table("metadata")) == [
        {"dataset": "constant", "version": "20220505.1", "format": "csv"}
    ]


def test_validate():
    assert True


@pytest.mark.parametrize(
    "source,expected",
    [
        (({}, {}), {}),
        (
            (
                {"a": {"ref": "map"}, "b": 2},
                {"map": {"values": {"1": True, "2": False}}},
            ),
            {"a": {"values": {"1": True, "2": False}}, "b": 2},
        ),
        (
            (
                {"a": [{"ref": "map"}, {"x": 4}]},
                {"map": {"values": {"1": True, "2": False}}},
            ),
            {"a": [{"values": {"1": True, "2": False}}, {"x": 4}]},
        ),
    ],
)
def test_expand_refs(source, expected):
    assert parser.expand_refs(*source) == expected


def test_reference_expansion():
    ps_noref = parser.Parser(TEST_PARSERS_PATH / "groupBy.json")
    ps_ref = parser.Parser(TEST_PARSERS_PATH / "groupBy-defs.json")
    del ps_ref.spec["adtl"]["defs"]
    assert ps_ref.spec == ps_noref.spec
