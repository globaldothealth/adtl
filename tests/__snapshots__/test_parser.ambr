# serializer version: 1
# name: test_default_date_format
  '''
  adtl_valid,adtl_error,epoch,id,some_date,text
  True,,1999-01-11,1,1999-01-24,Lorem ipsum
  True,,2022-12-19,2,2023-01-27,example
  
  '''
# ---
# name: test_multi_id_groupby
  '''
  admission_date,country_iso3,dataset_id,enrolment_date,sex_at_birth,subject_id
  2020-06-08,GBR,dataset-2020-03-23,2020-05-06,male,P007
  2020-06-08,GBR,dataset-2020-03-23,2022-01-11,female,S001
  
  '''
# ---
# name: test_parse_write_buffer
  '''
  admission_date,country_iso3,dataset_id,enrolment_date,sex_at_birth,subject_id
  2020-06-08,GBR,dataset-2020-03-23,2020-05-06,male,S007
  2020-06-08,GBR,dataset-2020-03-23,2022-01-11,female,S001
  
  '''
# ---
# name: test_validation
  '''
  adtl_valid,adtl_error,admission_date,country_iso3,dataset_id,enrolment_date,sex_at_birth,subject_id
  True,,2020-06-08,GBR,dataset-2020-03-23,2020-05-06,male,S007
  False,"data must contain ['subject_id', 'country_iso3', 'enrolment_date', 'sex_at_birth'] properties",8/6/2022,GBR,dataset-2020-03-23,2022-01-11,,S001
  False,data.admission_date must be date,8/6/2020,GBR,dataset-2020-03-23,2020-05-06,male,S009
  
  '''
# ---
